packages/ruleset/src/functions/property-inconsistent-name-and-type.js

```diff
@@ -0,0 +1,54 @@
+const { validateSubschemas } = require('../utils');
+
+let visitedProperties;
+let excludedProperties;
+
+module.exports = function(schema, options, { path }) {
+  visitedProperties = options.visitedProperties;
+  excludedProperties = options.excludedProperties;
+  return validateSubschemas(schema, path, propertyInconsistentNameAndType);
+};
+
+function propertyInconsistentNameAndType(schema, path) {
+  if (schema.properties) {
+    const errors = [];
+
+    for (const [propName, prop] of Object.entries(schema.properties)) {
+      // Skip check for deprecated properties.
+      if (prop.deprecated === true) continue;
+
+      if (visitedProperties[propName]) {
+        if (visitedProperties[propName].type !== prop.type) {
```

Comment:
`prop.type` is going to miss type for composed schemas, including a simple `allOf` reference, right? But if this is how our current rule behaves I suggest we leave it that way and open a separate issue to cover the more sophisticated behavior. (Especially since I have some helpful utilities for this in flight.)

Reply:
That's a good point. This is how the existing rule works but now you're making me realize that perhaps we shouldn't go about this rule this way. Instead of looking at the `properties` list, we could examine each schema as a whole. The `validateSubSchemas` function should give us all of the schemas in the API definition - we could just look to see what the type of each is and compare it to the other schemas we've seen.

Does this make sense? Would that address the issue you mentioned?